cylc/uiserver/data_mgr.py

```diff
@@ -13,81 +13,231 @@
+                for w_id in list(self.w_subs):
+                    if w_id not in self.workflows_mgr.workflows:
+                        self.w_subs[w_id].stop()
+                        del self.w_subs[w_id]
+                        if w_id in self.data:
+                            del self.data[w_id]
+
+                await asyncio.sleep(1.0)
+
+    def _start_subscription(self, w_id, host, port):
+        """Instatiate and run subscriber and data-store management.
+
+        Args:
+            w_id (str): Workflow external ID.
+            host (str): Hostname of target workflow.
+            port (int): Port of target workflow.
+
+        """
+        self.w_subs[w_id] = WorkflowSubscriber(
+            host, port,
+            context=self.workflows_mgr.context, topics=self.topics)
+        self.w_subs[w_id].loop.run_until_complete(
+            self.w_subs[w_id].subscribe(
+                process_delta_msg,
+                func=self.update_workflow_data,
+                w_id=w_id))
+
+    def update_workflow_data(self, topic, delta, w_id):
+        """Manage and apply incomming data-store deltas.
+
+        Args:
+            topic (str): topic of published data.
+            delta (object): Published protobuf message data container.
+            w_id (str): Workflow external ID.
+
+        """
+        # wait until data-store is populated for this workflow
+        loop_cnt = 0
+        while loop_cnt < 5:
+            if w_id not in self.data:
+                sleep(0.5)
+                loop_cnt += 1
+                continue
+            break
+        if w_id not in self.data:
+            return
+        if topic == 'shutdown':
+            self.workflow_mgr.stopping.add(w_id)
+            self.w_subs[w_id].stop()
+            return
+        delta_time = getattr(
+                delta, 'time', getattr(delta, 'last_updated', 0.0))
+        # If the workflow has reloaded recreate the data
+        # otherwise apply the delta if it's newer than the previously applied.
+        if delta.reloaded:
+            self.data[w_id][topic] = {ele.id: ele for ele in delta.deltas}
+            self.data[w_id]['delta_times'][topic] = delta_time
+        elif delta_time >= self.data[w_id]['delta_times'][topic]:
+            apply_delta(topic, delta, self.data[w_id])
+            self.data[w_id]['delta_times'][topic] = delta_time
+            self.reconcile_update(topic, delta, w_id)
+
+    def reconcile_update(self, topic, delta, w_id):
+        """Reconcile local with workflow data-store.
+
+        Verify data-store is in sync by topic/element-type
+        and on failure request entire set of respective data elements.
+
+        Args:
+            topic (str): topic of published data.
+            delta (object): Published protobuf message data container.
+            w_id (str): Workflow external ID.
+
+        """
+        if topic == WORKFLOW:
+            return
+        if topic == EDGES:
```

Comment:
Because edges don't have dynamic fields (no plan to change that (yet!)). Once created they don't change, a suite reload replaces existing, and they only get deleted.  (guess for consistency we could still do it the same way...)